### PR TITLE
Mix.install accepts atoms as paths

### LINKED ISSUE #12195: Allow atoms to be given to config_path and lockfile in Mix.install/2
Instead of this:

```elixir
my_app_root = Path.join(__DIR__, "..")

Mix.install(
  [
    {:my_app, path: my_app_root, env: :dev}
  ],
  config_path: Path.join(my_app_root, "config/config.exs"),
  lockfile: Path.join(my_app_root, "mix.lock")
)
```

We want this:

```elixir
Mix.install(
  [
    {:my_app, path: Path.join(__DIR__, ".."), env: :dev}
  ],
  config_path: :my_app
  lockfile: :my_app 
)
```

It will check there is a path dependency named `my_app` and automatically build its path.

### PR DESCRIPTION
**Pull request by sabiwara**
Close #12195

### CODE REVIEW

File: lib/mix/lib/mix.ex [hunk 1]
```diff
@@ -823,6 +834,13 @@ defmodule Mix do
     end
   end
 
+  defp expand_path(_path = nil, _deps, _relative_path), do: nil
+  defp expand_path(path, _deps, _relative_path) when is_binary(path), do: path
+
+  defp expand_path(app_name, deps, relative_path) when is_atom(app_name) do
+    deps |> Keyword.fetch!(app_name) |> Keyword.fetch!(:path) |> Path.join(relative_path)
```
Comment (josevalim) [inline lib/mix/lib/mix.ex:841]: Unfortunately path dependencies can be written in the format:

```elixir
{:dep, "~> 0.1.0", path: ...}
```

My suggestion would be something like this:

```elixir
case List.keyfind(deps, app_name, 0) do
  {_, _, opts} -> opts[:path] || raise ArgumentError, "#{app_name} given to :config_path must be a path dependency"
  {_, opts} -> opts[:path] || raise ArgumentError, "#{app_name} given to :config_path must be a path dependency"
  _ ->  raise ArgumentError, "unknown dependency #{app_name} given to :config_path"
end
```

Comment (sabiwara) [reply lib/mix/lib/mix.ex:841]: Oh sorry, I totally missed that 👍
Will fix.

Comment (sabiwara) [reply lib/mix/lib/mix.ex]: Done in [c23bdda](https://github.com/elixir-lang/elixir/pull/12204/commits/c23bdda44c740b080fa45cda9b8ab43a1eda1a79)

File: lib/mix/lib/mix.ex [hunk 2]
```diff
@@ -705,7 +716,7 @@ defmodule Mix do
       end)
 
     config = Keyword.get(opts, :config, [])
-    config_path = opts[:config_path] && Path.expand(opts[:config_path])
+    config_path = expand_path(opts[:config_path], deps, :config_path)
```
Comment (josevalim) [inline lib/mix/lib/mix.ex:719]: Tiny nitpick: I would pass the inner path here, so the function doesn't have to know the keys given to the caller:

```suggestion
    config_path = expand_path(opts[:config_path], deps, :config_path, "config/config.exs")
```

File: lib/mix/lib/mix.ex [hunk 3]
```diff
@@ -823,6 +834,30 @@ defmodule Mix do
     end
   end
 
+  defp expand_path(_path = nil, _deps, _key), do: nil
+  defp expand_path(path, _deps, _key) when is_binary(path), do: path
```
Comment (josevalim) [inline lib/mix/lib/mix.ex:838]:
```suggestion
  defp expand_path(path, _deps, _key) when is_binary(path), do: Path.expand(path)
```

File: lib/mix/lib/mix.ex (hunk 2 above)
Comment (sabiwara) [reply lib/mix/lib/mix.ex:719]: This was my original implementation, but we'll actually need it for the error message

Comment (josevalim) [reply lib/mix/lib/mix.ex:719]: I mean, pass both the key and the default value. :)

**Comment by josevalim** [comment]
:green_heart: :blue_heart: :purple_heart: :yellow_heart: :heart:

**Pull request merged**